lib/spack/spack/main.py

```diff
@@ -669,12 +669,7 @@ def main(argv=None):
         cmd_name = args.command[0]
         cmd_name = aliases.get(cmd_name, cmd_name)
 
-        try:
-            command = parser.add_command(cmd_name)
-        except ImportError:
-            if spack.config.get('config:debug'):
```

Comment:
With develop I get this in debug mode:
```console
$ spack -d foo
==> [2019-08-07-18:20:27.608459] Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2019-08-07-18:20:27.625975] Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/config.yaml
==> [2019-08-07-18:20:27.627948] Reading config file /home/mculpo/.spack/config.yaml
==> [2019-08-07-18:20:27.629310] Reading config file /home/mculpo/.spack/linux/config.yaml
Traceback (most recent call last):
  File "/home/mculpo/PycharmProjects/spack/bin/spack", line 48, in <module>
    sys.exit(spack.main.main())
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/main.py", line 673, in main
    command = parser.add_command(cmd_name)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/main.py", line 280, in add_command
    module = spack.cmd.get_module(cmd_name)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/cmd/__init__.py", line 97, in get_module
    level=0)
ImportError: No module named foo
```
with this branch I get:
```console
$ spack -d foo
==> [2019-08-07-18:20:00.939054] Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2019-08-07-18:20:00.959065] Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/config.yaml
==> [2019-08-07-18:20:00.960187] Reading config file /home/mculpo/.spack/config.yaml
==> [2019-08-07-18:20:00.961495] Reading config file /home/mculpo/.spack/linux/config.yaml
==> [2019-08-07-18:20:00.963405] CommandNotFoundError: foo is not a recognized Spack command or extension command.
    Known commands: activate add arch blame bootstrap build build-env buildcache cd checksum clean clone commands compiler compilers concretize config configure create deactivate debug dependencies dependents diy docs edit env extensions fetch filter find flake8 gpg graph help info install license list load location log-parse maintainers mirror module patch pkg providers pydoc python reindex release-jobs remove repo resource restage setup spec stage test uninstall unload unuse upload-s3 url use versions view
==> [2019-08-07-18:20:00.963496] Error: foo is not a recognized Spack command or extension command.
Known commands: activate add arch blame bootstrap build build-env buildcache cd checksum clean clone commands compiler compilers concretize config configure create deactivate debug dependencies dependents diy docs edit env extensions fetch filter find flake8 gpg graph help info install license list load location log-parse maintainers mirror module patch pkg providers pydoc python reindex release-jobs remove repo resource restage setup spec stage test uninstall unload unuse upload-s3 url use versions view
Traceback (most recent call last):
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/main.py", line 672, in main
    command = parser.add_command(cmd_name)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/main.py", line 280, in add_command
    module = spack.cmd.get_module(cmd_name)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/cmd/__init__.py", line 138, in get_module
    module = spack.extensions.load_command_extension(cmd_name)
  File "/home/mculpo/PycharmProjects/spack/lib/spack/spack/extensions.py", line 85, in load_command_extension
    raise CommandNotFoundError(command)
spack.extensions.CommandNotFoundError: foo is not a recognized Spack command or extension command.
    Known commands: activate add arch blame bootstrap build build-env buildcache cd checksum clean clone commands compiler compilers concretize config configure create deactivate debug dependencies dependents diy docs edit env extensions fetch filter find flake8 gpg graph help info install license list load location log-parse maintainers mirror module patch pkg providers pydoc python reindex release-jobs remove repo resource restage setup spec stage test uninstall unload unuse upload-s3 url use versions view
```
which is slightly more difficult to read (all the long lines show up as 4 lines on the screen)

Reply:
IIRC, the previous case did not distinguish between "command not found" and "error importing the found command," which was very frustrating during development. I can move the long line into `long_message` and see if that removes the duplication. The main motivation was to provide some context to allow the user to fix the problem in the case of _e.g._ misspellings.

Reply:
Upon checking, I already was putting the long context into `long_message` via the second argument to `SpackError.__init__()`, so this would seem to be more of a systemic issue. On the assumption that in the normal case the user is not using `--debug`, and this should give the non-developer the chance to fix their (likely) misspelled command without needing to resort to further debug, I would suggest keeping the longer message. I'd defer to you if you thought the message should be removed, but it would be nice if there were some way of avoiding the unnecessary duplication under `--debug` on a systemic basis.

Reply:
@alalazo @scheibelp With 5742d7a75, I have shortened the message to exhort the user to use `spack commands` to check the list of available commands rather than including them as part of the exception.  I think at least some of the unwanted duplication comes from the following clause in `main.py`:
```python
    except SpackError as e:
        tty.debug(e)
        e.die()  # gracefully die on any SpackErrors
```
I would suspect at the very least that the `tty.debug(e)` message could be removed. The next printout of the message comes (I think) from the call to `SpackError.print_context()` in `SpackError.die()`, and the final one comes from `sys.exit(1)` calling (presumably) `SpackError.__str__()`:
```python
    def die(self):
        self.print_context()
        sys.exit(1)
```
One wonders why `__str__()` unconditionally prints `long_message`: is that convention?
```python
    def __str__(self):
        msg = self.message
        if self._long_message:
            msg += "\n    %s" % self._long_message
        return msg
```